server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs

```diff
@@ -0,0 +1,389 @@
+  :: Client -> PayloadAction -> IO ()
+performPayloadAction client payloadAction =
+  case payloadAction of
+    ActionSubscribeWithAuth secret requestSubscribe
+      -> onPayloadSubscribeWithAuth client secret requestSubscribe
+    ActionSubscribeNoAuth requestSubscribe
+      -> onPayloadSubscribeNoAuth client requestSubscribe
+    ActionUnsubscribe requestUnsubscribe
+      -> onPayloadUnsubscribe client requestUnsubscribe
+    ActionError requestMalformed
+      -> onPayloadMalformed client requestMalformed
+
+onMessage :: Client -> IO ()
+onMessage client = do
+  let coreConfig = Core.coreConfig $ clientCore client
+      conn = clientConn client
+  dataMessage <- WS.receiveDataMessage conn
+  performPayloadAction client
+    $ determinePayloadAction coreConfig
+    $ parseDataMessage dataMessage
+
+onConnect :: Client -> IO ()
+onConnect client =
+  Core.withCoreMetrics (clientCore client) Metrics.incrementSubscribers
+
+onDisconnect :: Client -> IO ()
+onDisconnect client = do
+  let
+    core = clientCore client
+    subscriptions = clientSubscriptions client
+    uuid = clientUuid client
+
+  paths <-  HashMap.keys <$> MVar.takeMVar subscriptions
+  Monad.forM_ paths
+    (\path -> MVar.modifyMVar_ (Core.coreClients core)
+      (pure . Subscription.unsubscribe path uuid))
+
+  Core.withCoreMetrics core Metrics.decrementSubscribers
+
+handleClient :: WS.Connection -> Core -> IO ()
+handleClient conn core = do
+  uuid          <- newUUID
+  isDirty       <- MVar.newMVar ()
+  subscriptions <- MVar.newMVar (HashMap.empty :: HashMap [Text] (MVar Value))
+
+  let
+    client = Client
+      { clientConn = conn
+      , clientUuid = uuid
+      , clientCore = core
+      , clientIsDirty = isDirty
+      , clientSubscriptions = subscriptions
+      }
+
+    manageConnection = Async.withAsync
+      (updateThread client)
+      (const $ Monad.forever $ onMessage client)
+
+    -- Simply ignore connection errors, otherwise, Warp handles the exception
+    -- and sends a 500 response in the middle of a WebSocket connection, and
+    -- that violates the WebSocket protocol.
+    -- Note that subscribers are still properly removed by the finally below.
+    handleConnectionError :: WS.ConnectionException -> IO ()
+    handleConnectionError _ = pure ()
+  -- Put the client in the subscription tree and keep the connection open.
+  -- Remove it when the connection is closed.
+  Exception.finally (onConnect client >> manageConnection) (onDisconnect client)
+    `Exception.catch` handleConnectionError
+
+
+takeMVarUpdatedValues :: Client -> IO [(Text, Value)]
+takeMVarUpdatedValues client = do
+  let
+    isDirty = clientIsDirty client
+    subscriptions = clientSubscriptions client
+  MVar.takeMVar isDirty
+  valueMVars <- HashMap.toList <$> MVar.readMVar subscriptions
+  Maybe.catMaybes <$> (Monad.forM valueMVars $
+    \(path, valueMVar) ->
+      fmap (\v -> (Text.intercalate "/" path, v))
+      <$> MVar.tryTakeMVar valueMVar)
```

Comment:
I'm not sure what you mean. What are the consequences of the race condition?

I think the worst thing that could happen is that there is a false positive that the values are dirty, which would cause a read, but the values were already consumed. 

This will have no observable difference for the client. 

For the server, it wastes (possibly negligible)? CPU time.

It's indeed irritating that the consequences are unclear.

Reply:
Assuming that that's what the race condition is, the consequences are dependent on:
- How many path value MVars there are to traverse and do a tryTakeMVar on.
- How frequent the product of the subscribed paths become dirty.
  - Which itself depends on how many paths the thread is subscribed to.
  - Also depends on the nature of the icepeak data.

I feel like we can assume that a thread won't be subscribing to more than a dozen paths perhaps?
And for this race condition to occur often, it would mean that:
- The average time to traverse and do a `tryTakeMVar` on the  `n` MVars, is greater than the average period (1/frequency) of updates of the product of all paths.


I don't know if I'm making sense, but this is a fun thing to think about!

Reply:
@ErnestKz and I discussed this one previously. I think it should work as intended, with the only downside being the already mentioned extra traversal of the MVar's even though no more updates would be there.
I think the simplicity of this solution is worth that tradeoff, because we don't expect to have so many subscriptions that traversing all of them would be a performance problem.

Reply:
Sounds good!